view::remove_if(not_(&Class::method)) does not compile

Using:
- latest range-v3: commit 7baa4671ceafeaa8e97bb95b13e627e1704b1b59
- gcc 5.2.0 (official image from docker)

Test code:

```
#include <iostream>
#include <vector>

#include <range/v3/core.hpp>
#include <range/v3/utility/functional.hpp>
#include <range/v3/view/remove_if.hpp>

class Foo
{
private:
    int m_i;
public:
    Foo(int i) : m_i{i} {}
    int i() const { return m_i; }
    bool isOdd() const { return (m_i % 2) == 1; }
};

int main()
{
    const auto v1 = std::vector<Foo>{1,2,3,4,5,6};
    for (const auto& foo : v1) { std::cout << foo.i() << ','; } std::cout << '\n';
    const auto v2 = std::vector<Foo>(v1 | ranges::view::remove_if(ranges::not_(&Foo::isOdd)));
    //const auto v2 = std::vector<Foo>(v1 | ranges::view::remove_if(ranges::not_([](const auto& foo) { return foo.isOdd(); }))); // this compiles fine
    for (const auto& foo : v2) { std::cout << foo.i() << ','; } std::cout << '\n';
}
```

Output:

```
g++-5.2 -std=c++14 -O2 -march=native -Wall -Wextra -pedantic -I/home/etam/libs/range-v3/include/ test_cpp.cpp -o test_cpp
test_cpp.cpp: In function 'int main()':
test_cpp.cpp:57:93: error: invalid use of void expression
     const auto v2 = std::vector<Foo>(v1 | ranges::view::remove_if(ranges::not_(&Foo::isOdd)));
                                                                                             ^
test_cpp.cpp:59:28: error: unable to deduce 'auto&&' from 'v2'
     for (const auto& foo : v2) { std::cout << foo.i() << ','; } std::cout << '\n';
                            ^
In file included from /home/etam/libs/range-v3/include/range/v3/utility/box.hpp:23:0,
                 from /home/etam/libs/range-v3/include/range/v3/utility/functional.hpp:23,
                 from /home/etam/libs/range-v3/include/range/v3/utility/tagged_pair.hpp:21,
                 from /home/etam/libs/range-v3/include/range/v3/utility/dangling.hpp:21,
                 from /home/etam/libs/range-v3/include/range/v3/begin_end.hpp:23,
                 from /home/etam/libs/range-v3/include/range/v3/core.hpp:17,
                 from test_cpp.cpp:39:
/home/etam/libs/range-v3/include/range/v3/view/remove_if.hpp: In instantiation of 'void ranges::v3::view::remove_if_fn::operator()(Rng&&, Pred) const [with Rng = const std::vector<Foo>&; Pred = ranges::v3::logical_negate<bool (Foo::*)() const>; int _concept_requires_157 = 42; typename std::enable_if<((_concept_requires_157 == 43) || (! typename meta::v1::detail::_and_<ranges::v3::concepts::models<ranges::v3::concepts::InputRange, T>, typename meta::v1::detail::_and_<std::is_same<std::integer_sequence<bool, typename ranges::v3::concepts::models<ranges::v3::concepts::Readable, decltype (ranges::v3::{anonymous}::begin(declval<T&>()))>::type:: value>, std::integer_sequence<bool, (typename ranges::v3::concepts::models<ranges::v3::concepts::Readable, decltype (ranges::v3::{anonymous}::begin(declval<T&>()))>::type:: value || true)> >, meta::v1::defer<meta::v1::apply, meta::v1::compose<meta::v1::bind_front<meta::v1::quote<meta::v1::apply_list>, meta::v1::detail::on_<meta::v1::quote<meta::v1::fast_and>, meta::v1::bind_front<meta::v1::quote<meta::v1::apply_list>, meta::v1::bind_front<meta::v1::quote<ranges::v3::Predicate>, decltype (ranges::v3::{anonymous}::as_function(declval<F>()))> > > >, meta::v1::quote<ranges::v3::detail::iter_args_lists_> >, decltype (ranges::v3::{anonymous}::begin(declval<T&>()))> >::type>::type())), int>::type <anonymous> = 0]':
/usr/local/include/c++/5.2.0/functional:1074:55:   required from '_Result std::_Bind<_Functor(_Bound_args ...)>::__call(std::tuple<_Args2 ...>&&, std::_Index_tuple<_Indexes ...>) [with _Result = void; _Args = {const std::vector<Foo, std::allocator<Foo> >&}; long unsigned int ..._Indexes = {0ul, 1ul}; _Functor = ranges::v3::view::remove_if_fn; _Bound_args = {std::_Placeholder<1>, ranges::v3::logical_negate<bool (Foo::*)() const>}]'
/usr/local/include/c++/5.2.0/functional:1133:24:   required from '_Result std::_Bind<_Functor(_Bound_args ...)>::operator()(_Args&& ...) [with _Args = {const std::vector<Foo, std::allocator<Foo> >&}; _Result = void; _Functor = ranges::v3::view::remove_if_fn; _Bound_args = {std::_Placeholder<1>, ranges::v3::logical_negate<bool (Foo::*)() const>}]'
/home/etam/libs/range-v3/include/range/v3/view/view.hpp:93:17:   required from 'static decltype (v.view_(forward<Rng>(rng))) ranges::v3::view::view<View>::pipe(Rng&&, Vw&&) [with Rng = const std::vector<Foo>&; Vw = ranges::v3::view::view<ranges::v3::detail::pipeable_binder<std::_Bind<ranges::v3::view::remove_if_fn(std::_Placeholder<1>, ranges::v3::logical_negate<bool (Foo::*)() const>)> > >&; int _concept_requires_91 = 42; typename std::enable_if<((_concept_requires_91 == 43) || typename meta::v1::detail::_and_<typename meta::v1::detail::_and_<ranges::v3::concepts::models<ranges::v3::concepts::Range, Rng>, typename meta::v1::detail::_or_<std::is_lvalue_reference<Rng>, ranges::v3::concepts::models<ranges::v3::concepts::View, Container> >::type>::type, ranges::v3::concepts::models<ranges::v3::concepts::Function, View, Rng> >::type()), int>::type <anonymous> = 0; View = ranges::v3::detail::pipeable_binder<std::_Bind<ranges::v3::view::remove_if_fn(std::_Placeholder<1>, ranges::v3::logical_negate<bool (Foo::*)() const>)> >; decltype (v.view_(forward<Rng>(rng))) = void]'
/home/etam/libs/range-v3/include/range/v3/utility/functional.hpp:509:9:   required from 'decltype (ranges::v3::pipeable_access::impl<Pipe>::pipe(forward<Arg>(arg), pipe)) ranges::v3::operator|(Arg&&, Pipe) [with Arg = const std::vector<Foo>&; Pipe = ranges::v3::view::view<ranges::v3::detail::pipeable_binder<std::_Bind<ranges::v3::view::remove_if_fn(std::_Placeholder<1>, ranges::v3::logical_negate<bool (Foo::*)() const>)> > >; int _concept_requires_507 = 42; typename std::enable_if<((_concept_requires_507 == 43) || ((! ranges::v3::is_pipeable<T>()) && ranges::v3::is_pipeable<Pipe>())), int>::type <anonymous> = 0; decltype (ranges::v3::pipeable_access::impl<Pipe>::pipe(forward<Arg>(arg), pipe)) = void]'
test_cpp.cpp:57:92:   required from here
/home/etam/libs/range-v3/include/range/v3/utility/concepts.hpp:876:28: error: static assertion failed: The second argument to view::remove_if must be callable with a value of the range, and the return type must be convertible to bool
 #define CONCEPT_ASSERT_MSG static_assert
                            ^
/home/etam/libs/range-v3/include/range/v3/view/remove_if.hpp:163:21: note: in expansion of macro 'CONCEPT_ASSERT_MSG'
                     CONCEPT_ASSERT_MSG(IndirectCallablePredicate<Pred, range_iterator_t<Rng>>(),
                     ^
```

--- a/include/range/v3/utility/functional.hpp
+++ b/include/range/v3/utility/functional.hpp
@@ -197,40 +197,27 @@ namespace ranges
         struct logical_negate
         {
         private:
-            Pred pred_;
+            using fn_t = meta::_t<std::decay<function_type<Pred>>>;
+            fn_t pred_;
         public:
             logical_negate() = default;
 
             explicit constexpr logical_negate(Pred pred)
-              : pred_((Pred &&) pred)
+              : pred_(as_function((Pred &&) pred))
             {}
 
-            template<typename T,
-                CONCEPT_REQUIRES_(Predicate<Pred, T>())>
-            bool operator()(T && t)
+            template<typename ...Args,
+                CONCEPT_REQUIRES_(Predicate<fn_t&, Args...>())>
+            bool operator()(Args &&...args)
             {
-                return !pred_((T &&) t);
+                return !pred_(((Args &&) args)...);
             }
             /// \overload
-            template<typename T,
-                CONCEPT_REQUIRES_(Predicate<Pred const, T>())>
-            constexpr bool operator()(T && t) const
+            template<typename ...Args,
+                CONCEPT_REQUIRES_(Predicate<fn_t const&, Args...>())>
+            constexpr bool operator()(Args &&...args) const
             {
-                return !pred_((T &&) t);
-            }
-            /// \overload
-            template<typename T, typename U,
-                CONCEPT_REQUIRES_(Predicate<Pred, T, U>())>
-            bool operator()(T && t, U && u)
-            {
-                return !pred_((T &&) t, (U &&) u);
-            }
-            /// \overload
-            template<typename T, typename U,
-                CONCEPT_REQUIRES_(Predicate<Pred const, T, U>())>
-            constexpr bool operator()(T && t, U && u) const
-            {
-                return !pred_((T &&) t, (U &&) u);
+                return !pred_(((Args &&) args)...);
             }
         };
 

--- a/test/utility/functional.cpp
+++ b/test/utility/functional.cpp
@@ -10,7 +10,16 @@
 // Project home: https://github.com/ericniebler/range-v3
 
 #include <range/v3/utility/functional.hpp>
+#include <range/v3/view/filter.hpp>
 #include "../simple_test.hpp"
+#include "../test_utils.hpp"
+
+struct Integer
+{
+    int i;
+    operator int() const { return i; }
+    bool odd() const { return (i % 2) != 0; }
+};
 
 int main()
 {
@@ -50,5 +59,13 @@ int main()
     CHECK(&ri3_ == &ci);
 
     detail::ignore_unused(ri_, ri1_, ri2_, ri3_, j);
+
+    {
+      // Check that not_ works with callables
+      Integer some_ints[] = {{0}, {1}, {2}, {3}, {4}, {5}, {6}, {7}};
+      ::check_equal(some_ints | view::filter(ranges::not_(&Integer::odd)),
+                    {0,2,4,6});
+    }
+
     return test_result();
 }

